### ISSUE TITLE
Resolve as Won't Fix Issues not always syncing with Sonar Lint Connected Mode

### ISSUE THREAD

**Issue by Paulie-Waulie** [SonarSource/sonarlint-visualstudio#574]
### Description

This is similar to another issue I opened last year, however we are on later versions of Sonar, Sonar C# Analzyser now and the latest SonarLint. We have moved our sonar instance to a central one for the whole company (50+ teams), this instance is ultimately more up to date than the one we used for our office's five teams. The issue was resolved for the most part except I did not notice some problems with Won't Fix resolutions still being reported in Visual Studio at some times. However the problem now seems to be more persistent.

The problem reported last year was : [here](https://github.com/SonarSource/sonarlint-visualstudio/issues/445)

We have three code smells reported in Sonar Qube, two are genuine tech debt which we intend to resolve and a third which we marked as Resolved Won't Fix. When the analysis runs on the server only the two that we want reported are reported. In our IDE, when we perform a rebuild  the issue resolved as Won't Fix is reported in the Visual Studio error list. If I then open the file in question that warning is no longer reported unless I close the file and Rebuild again. Sometimes this does not occur, sometimes once the file has been opened any subsequent rebuilds do not cause the issue to re-appear but if we close the solution and re-open and build the issue always returns. The only way to get the issue to disappear again is to open the file.

Obviously we want to keep our IDE error list as clean as possible in terms of errors/warnings in the error list so to have the issues reported is adding extra noise, thankfully we only have this one issues marked as Won't Fix but if we had many more this would be a much bigger problem.

### Repro steps

1. Write code that violates a Sonar Rule to invoke an analysis build warning.

2. Execute Sonar Qube analysis to see the violation recorded as code smell in Sonar Qube.

3. Mark as Resolved Won't Fix.

4. Run analysis again to see the code smell no longer reported in Sonar Qube.

5. Perform a build/rebuild in Visual Studio. Issue still reported as build warning.

6. Open offending file, the issue is no longer reported.

7. Close file (and sometimes) close solution.

8. Perform a Rebuild and the issue is reported again.

9. Open offending file, the issue is no longer reported.

Note : Step 7 is intermittent, sometimes just closing the file causes the issue to be reported again, however closing and re-oping the solution always causes the issue to be reported again.

### Expected behavior

Resolved Marked as Won't Fix issues in Sonar Qube are no longer reported by SonarAnalyzer as long as the sync from SonarLint to the SonarQube instance has occurred which I according to "duncanp-sonar" in my previous reported issue occurs every ten minutes or following opening the solution.

### Actual behavior

Issues Marked as Won't Fix are reported in the IDE following a Rebuild unless the file is opened in the IDE.

### Known workarounds

Open the file.

### Related information

* SonarLint for VS Version 3.10.0.3095
* Visual Studio Version Enterprise 15.2 (26430.13)
* SonarQube server version Version 6.7 (build 33306)
* Languages in your solution - Just C#
* Sonar C# Analyzer Version 6.7.1

**Comment by Evangelink** [comment]
Hi @Paulie-Waulie,

When you go to the SonarQube tab of the `Team Explorer` pane. Could you see that you are correctly connected to your SonarQube instance? Could you check in the output window (selecting the `SonarLint` entry from the drop down) if you have any error/warning displayed related?

We are actually blocking the callback/analysis from roslyn for our rules until we have fetched the rules. So if they do appear on startup that means there was a problem fetching the list. Including closed issues, would you mind telling me (roughly) how many issues there are? We have a ticket about using some newer SonarQube API that would allow to improve performances when fetching the list (https://github.com/SonarSource/sonarlint-visualstudio/issues/421).

**Comment by Paulie-Waulie** [comment]
Hi Evangelink,

I can confirm that everything looks fine in the team explorer pane, in the output for SonarLint after opening the IDE, the solution and viewing the SonarQube team explorer pane I can see the following:

Checking for unbound projects.
No unbound projects were found.
Checking if SonarQube Quality profile has changed.
Quality Profile is up-to-date.
Connecting to SonarQube server: https://sonarqube.our.domain.com/
Connecting to SonarQube server: Validating credentials
Connecting to SonarQube server: Retrieving organizations
Connecting to SonarQube server: Detecting server plugins
   Discovered a supported plugin: Language: 'C#', Minimum version: '5.0'
Connecting to SonarQube server: Retrieving projects
Connecting to SonarQube server: Succeeded
Connecting to SonarQube server: Downloading server settings.

But the issue does not disappear until I open the file in question, around 3 seconds after opening that file the issue disappears in the error list. Nothing else is reported in the output window when this occurs.

In terms of issues listed in Sonar, we have:

42 Fixed - These were all due to test coverage which were resolved by exclusions in Sonar configuration for the project. These would not have been reported by SonarLint anyway as they are not part of the Analyzer.

2 Unresovled - These are the two issues that we want to fix and do want shown in the IDE as a prompt for the engineers to fix using the "good boy scout" mentality.

1 Won't Fix - This is the issue we are having problems with, it is the rule for lines longer than 200 characters, we decided that in this very specific case we did not want to fix the issue as it is a constant that is storing a regular expression and we felt that the potential solutions were not better than the actual issue. Generally though we want this rule to be fired.


Many Thanks

Paul

**Comment by Paulie-Waulie** [comment]
Hi,

Does anyone have any further information on this please? We are using this with more teams and projects and sadly people are complaining about this exact issue. We were waiting for the connected mode to sync resolutions before we rolled it out but this does not help much.

It means that our error Window is cluttered with issues that have been resolved in SonarQube and people do not want to have to open the files in question to remove the warnings.

This is happening with our team's current solution which has two projects, one is production and one is a test project. We currently have around 15 issues that are marked as "Resolved Won't Fix" in SonarQube.

Many Thanks

Paul

**Comment by duncanp-sonar** [comment]
Update: this issue is partially fixed by the linked commits and other bug fixes in release 4.7. However, the user experience is still poor, particularly in VS2017. The current position is as follows:

### When a solution is already open - delays in updating the list of issues (both VS2015 and VS2017)
When a connected solution is open SLVS polls the server for an updated list of issues every ten minutes, so there will be delay between an issue being suppressed on the server and the issue being hidden/shown in VS.
Even after the updated list of issues has been fetched, the list of issues shown in the _Error List_ will not change until VS triggers a re-analysis of the relevant code (e.g. when the file containing an issue is edited). There is not a supported way of programmatically triggering a re-analysis.
**Action:** check whether there is now a supported way to trigger re-analysis, and if not open a feature request against VS.

### When opening a connected solution - all issues are shown initially in VS2017
When a connected solution is open in VS2017, all issues including those that are suppressed in Sonar will appear in the error list (this can take a minute or two, as the full solution analysis is executed in a separate process _ServiceHub.RoslynCodeAnalysisService32.exe_).
When VS decides to trigger re-analysis (e.g. when a file is edited), suppressions for issues in the affected code will take effect so those issue will start to be taken into account, so suppressed issues will no longer be reported.

By and large, this issue does not appear in VS2015 because of a difference in the way analyzers are triggered (in VS2015 analyzers are run in-process, whereas in VS2017 they appear to be run both in- and out- of process). There are rarer circumstances when the full list of issues can appear in VS2015 when opening a solution but this happens much more rarely and appears to be a timing-related issue (e.g. VS triggers analysis before SLVS has established a connection with the Sonar server).
**Action:** investigate alternatives for suppressing issues e.g.
* handle suppressions in analyzers that are running out of process
* generate a suppressions file

**Comment by Geminior** [comment]
This issue is still a thing, and while suppressing reported violations locally is an option, it is far from optimal.
Is this issue still being actively worked on?

**Comment by iamrahul127** [comment]
I am surprised to see this is not fixed yet after 3 years. I am using VS2019 Enterprise, SonarQube '8.9.0.43852' Dev Edition & SonarLint 4.35. When I mark any issue in SQ server (In this case S1135) I can see "Number of suppressions found: 1" in SonarLint logs but when I open file which has this suppressed issue, issue again appears in the "Output" windows. 

```
Loaded settings from "C:\Users\mahulrah\AppData\Roaming\SonarLint for Visual Studio\settings.json".
Current VS version:
    Visual Studio Enterprise 2019
    16.8.31005.135
    16.8.5
Checking for suppressions...
Number of suppressions found: 1    => SonarLint is well aware of suppressed issue.
Connected to SonarQube '8.9.0.43852'.
Initializing the telemetry package...
Finished initializing the telemetry package...
Initializing the daemon package...
Finished initializing the daemon package.
[Open in IDE] Creating request listener...
[Open in IDE] Checking availability of port 64120
[Open in IDE] Request listener created successfully. Listening on port 64120.
Initializing the SonarLint package...
Connected mode detected.
Finishing initializing the SonarLint package
Initializing the notifications package...
Loading notifications settings...
Connected: checking for notifications
Finished initializing the notifications package
[Taint] Initializing taint issues synchronization package...
[Taint] Fetched 0 taint vulnerabilities.
[Taint] Finished initializing taint issues synchronization package.
Checking for suppressions...
```


If this functionality is not working for years what's the use of "Connected Mode". This is real bad user experience. Can someone please prioritize and fix this?

![image](https://user-images.githubusercontent.com/24877322/121768568-75f05500-cb7c-11eb-83c3-64feebd1710e.png)

**Comment by duncanp-sonar** [comment]
This should be fixed by #1005.
